### PR TITLE
Allow specifying number of nodes that can be unready during validation

### PR DESCRIPTION
**Pull request by serathius**
Implement equivalent of ALLOWED_NOTREADY_NODES from kubeup for kops. 

/cc @hakman

### CODE REVIEW

**Comment by Jefftree** [comment]
Test failure looks to be a doc generation thing `make gen-cli-docs`

File: pkg/validation/validate_cluster.go [hunk 1]
```diff
@@ -109,7 +111,7 @@ func hasPlaceHolderIP(host string) (string, error) {
 	return "", nil
 }
 
-func NewClusterValidator(cluster *kops.Cluster, cloud fi.Cloud, instanceGroupList *kops.InstanceGroupList, filterInstanceGroups func(ig *kops.InstanceGroup) bool, filterPodsForValidation func(pod *v1.Pod) bool, restConfig *rest.Config, k8sClient kubernetes.Interface) (ClusterValidator, error) {
```
Comment (Jefftree) [inline pkg/validation/validate_cluster.go:112]: should we validate 0 <= percentage <= 100?

File: cmd/kops/validate_cluster.go [hunk 2]
```diff
@@ -69,6 +69,8 @@ type ValidateClusterOptions struct {
 	interval           time.Duration
 	kubeconfig         string
 
+	AllowedNotReadyNodesPercentage int
```
Comment (Jefftree) [inline cmd/kops/validate_cluster.go]: I think most of the failure have been just one node. Do we want more granularity or is 1% = 5 nodes fine?

**Comment by dims** [comment]
seems legit ci failure?

```
FAIL: make verify-gendocs failed, as the generated markdown docs are out of date.
FAIL: Please run the following command: make gen-cli-docs.
32,40c32,39
<       --allowed-not-ready-nodes-percentage int   The percentage of non-ready worker nodes tolerated during validation
<       --api-server string                        Override the API server used when communicating with the cluster kube-apiserver
<       --count int                                Number of consecutive successful validations required
<   -h, --help                                     help for cluster
<       --interval duration                        Time in duration to wait between validation attempts (default 10s)
<       --kubeconfig string                        Path to the kubeconfig file
<   -o, --output string                            Output format. One of json|yaml|table. (default "table")
<       --use-kubeconfig                           Use the server endpoint from the local kubeconfig instead of inferring from cluster name
<       --wait duration                            Amount of time to wait for the cluster to become ready
---
>       --api-server string   Override the API server used when communicating with the cluster kube-apiserver
>       --count int           Number of consecutive successful validations required
>   -h, --help                help for cluster
>       --interval duration   Time in duration to wait between validation attempts (default 10s)
>       --kubeconfig string   Path to the kubeconfig file
>   -o, --output string       Output format. One of json|yaml|table. (default "table")
>       --use-kubeconfig      Use the server endpoint from the local kubeconfig instead of inferring from cluster name
>       --wait duration       Amount of time to wait for the cluster to become ready
make: *** [Makefile:368: verify-gendocs] Error 1
```

File: cmd/kops/validate_cluster.go [hunk 3]
```diff
@@ -117,6 +120,7 @@ func NewCmdValidateCluster(f *util.Factory, out io.Writer) *cobra.Command {
 	cmd.Flags().IntVar(&options.count, "count", options.count, "Number of consecutive successful validations required")
 	cmd.Flags().DurationVar(&options.interval, "interval", options.interval, "Time in duration to wait between validation attempts")
 	cmd.Flags().StringVar(&options.kubeconfig, "kubeconfig", "", "Path to the kubeconfig file")
+	cmd.Flags().IntVar(&options.AllowedNotReadyNodesPercentage, "allowed-not-ready-nodes-percentage", options.AllowedNotReadyNodesPercentage, "The percentage of non-ready worker nodes tolerated during validation")
```
Comment (hakman) [inline cmd/kops/validate_cluster.go]:
```suggestion
	cmd.Flags().IntVar(&options.AllowedNotReadyNodesPercentage, "max-unready-nodes", options.AllowedNotReadyNodesPercentage, "Percentage of unready worker nodes tolerated during validation")
```

File: pkg/validation/validate_cluster.go [hunk 4]
```diff
@@ -73,6 +73,8 @@ type clusterValidatorImpl struct {
 
 	// filterPodsForValidation is a function that returns true if the pod should be validated
 	filterPodsForValidation func(pod *v1.Pod) bool
+
+	allowedNotReadyNodesPercentage int
```
Comment (hakman) [inline pkg/validation/validate_cluster.go]: Would it be possible to make this intOrString, to allow both percentage or number?

**Review by hakman — CHANGES_REQUESTED** [review]
I flagged some small issues, mostly consistency related.
Additionally, there might be an issue with counting the total nodes, for percentages. I think instead of worker nodes, all instance groups are taken into consideration.

File: pkg/validation/validate_cluster.go (hunk 4 above)
Comment (hakman) [reply pkg/validation/validate_cluster.go]: Or can stick to fixed number instead of percentage and simplify all this.

Comment (Jefftree) [reply pkg/validation/validate_cluster.go]: Would prefer using just fixed number since that's what the old script used and we don't need to go through the intorstring complication.

**Comment by serathius** [comment]
Addressed comments PTAL @hakman @Jefftree

**Comment by serathius** [comment]
/cc @upodroid

Disclosure, I vibecoded this and didn't test it manually. I have read it enough to say it's not horrible, but I don't know the kops code to be 100% sure there is some edge case missed. If you have any suggestion about how to test it locally (without needing cloud access) or something seems fishy please point it out early. I read it and it make sense, but I didn't have time to familiarize myself with kops code.

File: pkg/validation/validate_cluster.go [hunk 5]
```diff
@@ -254,6 +295,11 @@ func (v *ValidationCluster) collectPodFailures(ctx context.Context, client kuber
 			podNode = nodeInstanceGroupMapping[nodeByAddress[pod.Status.HostIP]]
 		}
 
+		nodeName := nodeByAddress[pod.Status.HostIP]
+		if toleratedNodes[nodeName] {
+			return nil
+		}
```
Comment (hakman) [inline pkg/validation/validate_cluster.go:301]: `Status.HostIP` may still be empty for a Pending pod already assigned to a NotReady node. Can we use the canonical `Spec.NodeName` relationship throughout and remove the address map?

```diff
diff --git a/pkg/validation/validate_cluster.go b/pkg/validation/validate_cluster.go
index 794e9f172a..fb0f4914c2 100644
--- a/pkg/validation/validate_cluster.go
+++ b/pkg/validation/validate_cluster.go
@@ -231,14 +231,7 @@ func (v *clusterValidatorImpl) Validate(ctx context.Context) (*ValidationCluster
 
 	readyNodes, nodeInstanceGroupMapping := validation.validateNodes(cloudGroups, v.allInstanceGroups, v.filterInstanceGroups, toleratedNodes)
 
-	nodeByAddress := map[string]string{}
-	for _, node := range nodeList.Items {
-		for _, nodeAddress := range node.Status.Addresses {
-			nodeByAddress[nodeAddress.Address] = node.Name
-		}
-	}
-
-	if err := validation.collectPodFailures(ctx, v.k8sClient, readyNodes, nodeByAddress, nodeInstanceGroupMapping, v.filterPodsForValidation, toleratedNodes); err != nil {
+	if err := validation.collectPodFailures(ctx, v.k8sClient, readyNodes, nodeInstanceGroupMapping, v.filterPodsForValidation, toleratedNodes); err != nil {
 		return nil, fmt.Errorf("cannot get pod health for %q: %v", v.cluster.Name, err)
 	}
 
@@ -251,7 +244,7 @@ var masterStaticPods = []string{
 	"kube-scheduler",
 }
 
-func (v *ValidationCluster) collectPodFailures(ctx context.Context, client kubernetes.Interface, readyNodes []v1.Node, nodeByAddress map[string]string, nodeInstanceGroupMapping map[string]*kops.InstanceGroup, podValidationFilter func(pod *v1.Pod) bool, toleratedNodes map[string]bool) error {
+func (v *ValidationCluster) collectPodFailures(ctx context.Context, client kubernetes.Interface, readyNodes []v1.Node, nodeInstanceGroupMapping map[string]*kops.InstanceGroup, podValidationFilter func(pod *v1.Pod) bool, toleratedNodes map[string]bool) error {
 	log := klog.FromContext(ctx)
 
 	masterWithoutPod := map[string]map[string]bool{}
@@ -272,8 +265,8 @@ func (v *ValidationCluster) collectPodFailures(ctx context.Context, client kuber
 		pod := obj.(*v1.Pod)
 
 		app := pod.GetLabels()["k8s-app"]
-		if pod.Namespace == "kube-system" && masterWithoutPod[nodeByAddress[pod.Status.HostIP]][app] {
-			delete(masterWithoutPod[nodeByAddress[pod.Status.HostIP]], app)
+		if pod.Namespace == "kube-system" && masterWithoutPod[pod.Spec.NodeName][app] {
+			delete(masterWithoutPod[pod.Spec.NodeName], app)
 		}
 
 		// Ignore pods that we don't want to validate
@@ -292,11 +285,10 @@ func (v *ValidationCluster) collectPodFailures(ctx context.Context, client kuber
 
 		var podNode *kops.InstanceGroup
 		if priority == "system-node-critical" {
-			podNode = nodeInstanceGroupMapping[nodeByAddress[pod.Status.HostIP]]
+			podNode = nodeInstanceGroupMapping[pod.Spec.NodeName]
 		}
 
-		nodeName := nodeByAddress[pod.Status.HostIP]
-		if toleratedNodes[nodeName] {
+		if toleratedNodes[pod.Spec.NodeName] {
 			return nil
 		}
 
```

**Review by hakman — CHANGES_REQUESTED** [review]

**Comment by hakman** [comment]
@serathius please also update the PR title and commit message to match the new scope.

**Comment by Jefftree** [comment]
Taking over this PR https://github.com/kubernetes/kops/pull/18614

**Comment by serathius** [comment]
Won't have time to address comments, thanks @Jefftree for https://github.com/kubernetes/kops/pull/18614 

@hakman please review https://github.com/kubernetes/kops/pull/18614
